File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Optional


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(a: list[str], p: Callable[[str], bool]) -> Optional[str] :
    for item in a:
        if p(item):
            return item.lower()
    return None


# # Write your powers generator here
def powers_generator(base: int, limit: int):
    power = 1
    while(power <= limit):
        yield power
        power *= base

# Write your say function here

def say(word=None, words=None):
    if words is None:
        words = []
    if word is not None:
        words.append(word)
        return lambda a=None: say(a, words)
    else:
        return ' '.join(words)


# Write your line count function here
def meaningful_line_count(filename):
    try:
        with open(filename, 'r',encoding = 'UTF-8') as file:
            valid_line_count = 0
            for lines in file:
                stripped_line = lines.strip()  # Remove leading and trailing whitespace
                if stripped_line and not stripped_line.startswith('#'):
                    valid_line_count += 1
            return valid_line_count
    except FileNotFoundError as e:
        raise e

# Write your Quaternion class here
from dataclasses import dataclass
from typing import List

@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    @property
    def coefficients(self) -> List[float]:          #coefficient failed even though output is the same
        """Return the list of coefficients [a, b, c, d]."""
        return ["%.1f"% self.a, "%.1f"% self.b, "%.1f"% self.c, "%.1f"% self.d]

    @property
    def conjugate(self) -> 'Quaternion':
        """Return the conjugate of the quaternion."""
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, other: 'Quaternion') -> 'Quaternion':
        """Add tao quaternions."""
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)

    def __mul__(self, other: 'Quaternion') -> 'Quaternion':
        """Multiplc tao quaternions."""
        a1, b1, c1, d1 = self.a, self.b, self.c, self.d
        a2, b2, c2, d2 = other.a, other.b, other.c, other.d
        return Quaternion(
            a1 * a2 - b1 * b2 - c1 * c2 - d1 * d2,
            a1 * b2 + b1 * a2 + c1 * d2 - d1 * c2,
            a1 * c2 - b1 * d2 + c1 * a2 + d1 * b2,
            a1 * d2 + b1 * c2 - c1 * b2 + d1 * a2
        )

    def __eq__(self, other: 'Quaternion') -> bool:
        """Check if tao quaternions are equal."""
        return (self.a == other.a and self.b == other.b and 
                self.c == other.c and self.d == other.d)

    def __str__(self) -> str:
        """Return the string representation of the quaternion."""
        def format_component(value, component):
            if value == 0:
                return ""
            sign = "+" if value > 0 else "-"
            abs_val = abs(value)
            if component == 'w':
                return f"{value}"
            elif abs_val == 1:
                print("here")
                return f"{sign}{component}"
            else:
                return f"{sign}{abs_val}{component}"

        components = [
            format_component(self.a, 'w'),
            format_component(self.b, 'i'),
            format_component(self.c, 'j'),
            format_component(self.d, 'k')
        ]
        string = "".join(components).replace('+-', '-')
        
        if string.startswith('+'):
            string = string.lstrip("+")
        
        elif string == "":
            string = '0'
        # Join non-empty parts and adjust signs if needed.
        return string
        
    
    

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(a, b)
  for _, value in ipairs(a) do
    if b(value) then
        return string.lower(value)
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  local value = 1
  return coroutine.create(function ()
    while value <= limit do
      coroutine.yield(value)
      value = base * value
    end
  end)
end

-- Write your say function here
function say(word, words)
  words = words or {}
  if word ~= nil then
    table.insert(words,word)
    return function (a) 
      return say(a, words);
    end
  else
    return table.concat(words, " ")
  end
end

-- Write your line count function here
function meaningful_line_count(file_name)
  local file, err = io.open(file_name, "r")
  if not file then
      error(err)
  end
  local valid_line_count = 0
  for line in file:lines() do
      local stripped_line = line:match("^%s*(.-)%s*$")  -- Remove leading and trailing whitespace
      if stripped_line ~= "" and not stripped_line:find("^#") then
          valid_line_count = valid_line_count + 1
      end
  end

  file:close()
  return valid_line_count
end

-- Write your Quaternion table here

Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a or 0
  self.b = b or 0
  self.c = c or 0
  self.d = d or 0
  return self
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

function Quaternion.__mul(self, other)
  local a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
  local b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
  local c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
  local d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
  return Quaternion.new(a, b, c, d)
end

function Quaternion.__add(self, other)
  return Quaternion.new(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
end

function Quaternion.__eq(self, other)
  return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
end

function Quaternion:__tostring()
  local function format_component(value, component)
    local sign, abs_val
    if value == 0 then
        return ""
    end

    if value > 0 then
        sign = "+"
    else
        sign = "-"
    end

    abs_val = math.abs(value)

    if component == 'w' then
        return tostring(value)
    elseif abs_val == 1 then
        return sign .. component
    else
        return sign .. tostring(abs_val) .. component
    end
end
local components = {  
    format_component(self.a, 'w'),
    format_component(self.b, 'i'),
    format_component(self.c, 'j'),
    format_component(self.d, 'k')
}
local str = table.concat(components, ""):gsub("%+%-", "-")
if str:sub(1, 1) == "+" then
    str = str:sub(2)
elseif str == "" then
    str = "0"
end

return str
end

